tests/bluetooth/tester/src/btp_vcp.c

```diff
@@ -0,0 +1,544 @@
+/* vcp.c - Bluetooth VCP Tester */
+
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/bluetooth/bluetooth.h>
+
+#include <zephyr/bluetooth/testing.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/bluetooth/audio/vcp.h>
+#include <zephyr/bluetooth/audio/aics.h>
+#include "zephyr/bluetooth/audio/vocs.h"
+#include <app_keys.h>
+
+#include <zephyr/logging/log.h>
+#define LOG_MODULE_NAME bttester_vcp
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include "btp/btp.h"
+
+#define CONTROLLER_INDEX 0
+
+struct bt_vcp_vol_rend_register_param vcp_register_param;
+struct bt_vcp_included included;
+static bool g_aics_active = 1;
+
+static void set_register_params(uint8_t gain_mode);
+
+/* Volume Control Service */
+static void vcs_supported_commands(uint8_t *data, uint16_t len)
+{
+	struct net_buf_simple *buf = NET_BUF_SIMPLE(BTP_DATA_MAX_SIZE);
+
+	net_buf_simple_init(buf, 0);
+	net_buf_simple_add_u8(buf, VCS_READ_SUPPORTED_COMMANDS);
+	net_buf_simple_add_u8(buf, VCS_INIT);
+	net_buf_simple_add_u8(buf, VCS_SET_VOL);
+	net_buf_simple_add_u8(buf, VCS_VOL_UP);
+	net_buf_simple_add_u8(buf, VCS_VOL_DOWN);
+	net_buf_simple_add_u8(buf, VCS_MUTE);
+	net_buf_simple_add_u8(buf, VCS_UNMUTE);
+
+	tester_send(BTP_SERVICE_ID_VCS, VCS_READ_SUPPORTED_COMMANDS,
+		    CONTROLLER_INDEX, buf->data, buf->len);
+}
+
+static void set_volume(uint8_t *data, uint16_t len)
+{
+	const struct vcs_set_vol_cmd *cmd = (void *) data;
+	uint8_t status = BTP_STATUS_SUCCESS;
+	uint32_t volume;
+	int err;
+
+	volume = sys_le32_to_cpu(cmd->volume);
+
+	LOG_DBG("volume 0x%04x", volume);
+
+	err = bt_vcp_vol_rend_set_vol(volume);
+	if (err) {
+		status = BTP_STATUS_FAILED;
+	}
+
+	tester_rsp(BTP_SERVICE_ID_VCS, VCS_SET_VOL, CONTROLLER_INDEX, status);
+}
+
+static void vol_up(void)
+{
+	uint8_t status = BTP_STATUS_SUCCESS;
+	int err;
+
+	LOG_DBG("Volume Up");
+
+	err = bt_vcp_vol_rend_vol_up();
+	if (err) {
+		status = BTP_STATUS_FAILED;
+	}
+
+	tester_rsp(BTP_SERVICE_ID_VCS, VCS_VOL_UP, CONTROLLER_INDEX, status);
+}
+
+static void vol_down(void)
+{
+	uint8_t status = BTP_STATUS_SUCCESS;
+	int err;
+
+	LOG_DBG("Volume Down");
+
+	err = bt_vcp_vol_rend_vol_down();
+	if (err) {
+		status = BTP_STATUS_FAILED;
+	}
+
+	tester_rsp(BTP_SERVICE_ID_VCS, VCS_VOL_DOWN, CONTROLLER_INDEX, status);
+}
+
+static void mute(void)
+{
+	uint8_t status = BTP_STATUS_SUCCESS;
+	int err;
+
+	err = bt_vcp_vol_rend_mute();
+	if (err) {
+		status = BTP_STATUS_FAILED;
+	}
+
+	tester_rsp(BTP_SERVICE_ID_VCS, VCS_MUTE, CONTROLLER_INDEX, status);
+}
+
+static void unmute(void)
+{
+	uint8_t status = BTP_STATUS_SUCCESS;
+	int err;
+
+	err = bt_vcp_vol_rend_unmute();
+	if (err) {
+		status = BTP_STATUS_FAILED;
+	}
+
+	tester_rsp(BTP_SERVICE_ID_VCS, VCS_UNMUTE, CONTROLLER_INDEX, status);
+}
```

Comment:
What is the rationale behind having `unmute` calling `tester_rsp`, rather than having `tester_handle_vcs` call that, since `tester_handle_vcs` could simply do something like

```c
uint8_t status;

...
	case VCS_MUTE:
		status= mute();
		break;
	case VCS_UNMUTE:
		status = unmute();
		break;

tester_rsp(BTP_SERVICE_ID_VCS, opcode, index, status);
```

In fact, whatever is calling `tester_handle_vcs` could do the same, and then we could have just a single

```c
tester_rsp(service_id, opcode, index, status);
```

if `tester_handle_vcs` returns the `status`, instead of one for each service_id and opcode, where the values are hardcoded (which is quite an error prone approach)

Reply:
This is due to btp design, some test cases will require to return more than just `status` that's why this is generally handled by functions nested in those switch case structures. Yes in my case commands only return `status` but in the future, there might be more demanding test cases :).

Reply:
Do we have anything right now that needs to return more than a `status` field? Would be nice to have some code to reference for that argument

Reply:
Right now no, but I did it like this to be similar to other BTP services.

Reply:
That's fair; something we can consider doing across all the commands then; would probably clean the code up a fair bit and generally less error prone